api/layer/multipart_upload.go

```diff
@@ -0,0 +1,494 @@
+		{attr: UploadIDAttributeName, val: p.UploadID},
+		{attr: UploadPartNumberAttributeName, val: "0"},
+		{attr: UploadKeyAttributeName, val: p.Key},
+	}
+	f := &findParams{
+		filters: filters,
+		cid:     p.Bkt.CID,
+	}
+	ids, err := n.objectSearch(ctx, f)
+	if err != nil {
+		if errors.IsS3Error(err, errors.ErrNoSuchKey) {
+			return nil, errors.GetAPIError(errors.ErrNoSuchUpload)
+		}
+		return nil, err
+	}
+	if ids == nil {
+		return nil, errors.GetAPIError(errors.ErrNoSuchUpload)
+	}
+	meta, err := n.objectHead(ctx, p.Bkt.CID, ids[0])
+	if err != nil {
+		return nil, err
+	}
+	return objInfoFromMeta(p.Bkt, meta), nil
+
+}
+
+func (n *layer) getUploadParts(ctx context.Context, uploadID, key string, bktInfo *data.BucketInfo) (map[int]*data.ObjectInfo, error) {
+	f := &findParams{
+		filters: []*filter{
+			{attr: UploadIDAttributeName, val: uploadID},
+			{attr: UploadKeyAttributeName, val: key}},
+		cid: bktInfo.CID,
+	}
+
+	ids, err := n.objectSearch(ctx, f)
+	if err != nil {
+		if errors.IsS3Error(err, errors.ErrNoSuchKey) {
+			return nil, errors.GetAPIError(errors.ErrNoSuchUpload)
+		}
+		return nil, err
+	}
+
+	res := make(map[int]*data.ObjectInfo)
+
+	for _, id := range ids {
+		meta, err := n.objectHead(ctx, bktInfo.CID, id)
+		if err != nil {
+			n.log.Warn("couldn't head a part of upload",
+				zap.Stringer("object id", id),
+				zap.Stringer("bucket id", bktInfo.CID),
+				zap.Error(err))
+			continue
+		}
+		info := objInfoFromMeta(bktInfo, meta)
+		numStr := info.Headers[UploadPartNumberAttributeName]
+		num, err := strconv.Atoi(numStr)
+		if err != nil {
+			return nil, errors.GetAPIError(errors.ErrInternalError)
+		}
+		res[num] = info
+	}
+
+	return res, nil
+}
+
+func createUploadPartName(uploadId, key string, partNumber int) string {
+	return UploadPartKeyPrefix + uploadId + "-" + key + "-" + strconv.Itoa(partNumber)
+}
+
+func trimAfterUploadIDAndKey(key, id string, uploads []*UploadInfo) []*UploadInfo {
+	var res []*UploadInfo
+	if len(uploads) != 0 && uploads[len(uploads)-1].Key < key {
+		return res
+	}
+
+	for _, obj := range uploads {
+		if obj.Key >= key && obj.UploadID > id {
+			res = append(res, obj)
+		}
+	}
+
+	return res
+}
+
+func trimAfterUploadKey(key string, objects []*UploadInfo) []*UploadInfo {
+	if len(objects) != 0 && objects[len(objects)-1].Key <= key {
+		return []*UploadInfo{}
+	}
+	for i, obj := range objects {
+		if obj.Key > key {
+			return objects[i:]
+		}
+	}
+
+	return []*UploadInfo{}
+}
+
+func uploadInfoFromMeta(meta *object.Object, prefix, delimiter string) *UploadInfo {
+	var (
+		isDir       bool
+		creation    time.Time
+		userHeaders = userHeaders(meta.Attributes())
+		key         = userHeaders[UploadKeyAttributeName]
+	)
+
+	if !strings.HasPrefix(key, prefix) {
+		return nil
+	}
+
+	if val, ok := userHeaders[object.AttributeTimestamp]; !ok {
+		// ignore empty value
+	} else if dt, err := strconv.ParseInt(val, 10, 64); err == nil {
+		creation = time.Unix(dt, 0)
+	}
```

Comment:
We can remove empty branch
```suggestion
	if val, ok := userHeaders[object.AttributeTimestamp]; ok {
		if dt, err := strconv.ParseInt(val, 10, 64); err == nil {
			creation = time.Unix(dt, 0)
		}
	}
```